nsfc-plugin/src/main/java/org/osc/controller/nsfc/api/NeutronSfcSdnRedirectionApi.java

```diff
@@ -87,25 +167,139 @@ public void removeInspectionPort(InspectionPortElement inspectionPort)
+        } else {
+            LOG.warn("Attempt to remove nonexistent Inspection Port for ingress {} and egress {}",
+                    inspectionPort.getIngressPort(), inspectionPort.getEgressPort());
+        }
     }
 
     // Inspection Hooks methods
     @Override
-    public String installInspectionHook(NetworkElement inspectedPort, InspectionPortElement inspectionPort, Long tag,
-            TagEncapsulationType encType, Long order, FailurePolicyType failurePolicyType)
+    public String installInspectionHook(NetworkElement inspectedPortElement,
+                                        InspectionPortElement inspectionPortElement, Long tag,
+                                        TagEncapsulationType encType, Long order,
+                                        FailurePolicyType failurePolicyType)
             throws NetworkPortNotFoundException, Exception {
 
-        return null; // TODO (Dmitry) Implement
+        checkArgument(inspectedPortElement != null && inspectedPortElement.getElementId() != null,
+                      "null passed for %s !", "Service Function Chain");
+        checkArgument(inspectionPortElement != null && inspectionPortElement.getElementId() != null,
+                      "null passed for %s !", "Service Function Chain");
+
+        if (inspectedPortElement.getPortIPs() == null || inspectedPortElement.getPortIPs().size() == 0) {
+            String msg = String.format("Inspected port %s has no address to protect!", inspectedPortElement.getElementId());
+            LOG.error(msg);
+            throw new IllegalStateException(msg);
+        }
+
+        LOG.info("Installing Inspection Hook for (Inspected Port {} ; Inspection Port {}):",
+                inspectedPortElement, inspectionPortElement);
+
+        PortChain portChain = this.osCalls.getPortChain(inspectionPortElement.getElementId());
+        checkArgument(portChain != null && portChain.getId() != null,
+                      "Cannot find %s by id: %s!", "Service Function Chain", inspectionPortElement.getElementId());
+        checkArgument(portChain.getPortPairGroups() != null && portChain.getPortPairGroups().size() > 0,
+                      "Cannot install inspection hook with empty port chain!");
+
+        // if inspectedPort is being protected, it doesn't matter what is the inspection port
+        FlowClassifier flowClassifier = this.utils.fetchInspHookByInspectedPort(inspectedPortElement);
+        if (flowClassifier != null) {
+            String msg = String.format("Found existing inspection hook (Inspected %s ; Inspection Port %s)",
+                    inspectedPortElement, flowClassifier.getId());
+            LOG.error(msg + " " + flowClassifier);
+            throw new IllegalStateException(msg);
+        }
+
+        ServiceFunctionChainEntity sfcEntity = this.utils.fetchSFCWithAllDepends(portChain);
+
+        String inspectedIp = inspectedPortElement.getPortIPs().get(0);
```

Comment:
not ips...we protect by logical ports. Might be good to sync with karim to get more details on this.

Some of the comments will become redundant/clear if the code is tested with real openstack environments. Please make sure to test with deployed openstack.

Reply:
The code in OSGiIntegrationTest actually works with my openstsack environment. I just disable it for Travis, it won't build there.